File: 2023/day09/node/solution.js
import { readFileSync } from 'fs';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';

const __dirname = dirname(fileURLToPath(import.meta.url));
const input = readFileSync(join(__dirname, '..', 'input.txt'), 'utf-8').trim();

function parseInput(text) {
    return text.split('\n').map(line =>
        line.split(/\s+/).map(Number)
    );
}

function getDifferences(seq) {
    const result = [];
    for (let i = 0; i < seq.length - 1; i++) {
        result.push(seq[i + 1] - seq[i]);
    }
    return result;
}

function extrapolateNext(seq) {
    const sequences = [[...seq]];
    let current = seq;

    while (!current.every(x => x === 0)) {
        current = getDifferences(current);
        sequences.push(current);
    }

    for (let i = sequences.length - 2; i >= 0; i--) {
        sequences[i].push(sequences[i].at(-1) + sequences[i + 1].at(-1));
    }

    return sequences[0].at(-1);
}

function extrapolatePrev(seq) {
    const sequences = [[...seq]];
    let current = seq;

    while (!current.every(x => x === 0)) {
        current = getDifferences(current);
        sequences.push(current);
    }

    for (let i = sequences.length - 2; i >= 0; i--) {
        sequences[i].unshift(sequences[i][0] - sequences[i + 1][0]);
    }

    return sequences[0][0];
}

const histories = parseInput(input);

function part1() {
    return histories.reduce((sum, h) => sum + extrapolateNext(h), 0);
}

function part2() {
    return histories.reduce((sum, h) => sum + extrapolatePrev(h), 0);
}

console.log('Part 1:', part1());
console.log('Part 2:', part2());

File: 2023/day09/java/Solution.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Solution {

    public static void main(String[] args) throws IOException {
        String input = Files.readString(Path.of("../input.txt")).trim();
        List<long[]> histories = parseInput(input);

        System.out.println("Part 1: " + part1(histories));
        System.out.println("Part 2: " + part2(histories));
    }

    private static List<long[]> parseInput(String text) {
        return Arrays.stream(text.split("\n"))
                .map(line -> Arrays.stream(line.split("\\s+"))
                        .mapToLong(Long::parseLong)
                        .toArray())
                .toList();
    }

    private static long[] getDifferences(long[] seq) {
        long[] diffs = new long[seq.length - 1];
        for (int i = 0; i < seq.length - 1; i++) {
            diffs[i] = seq[i + 1] - seq[i];
        }
        return diffs;
    }

    private static boolean allZeros(long[] seq) {
        for (long val : seq) {
            if (val != 0) return false;
        }
        return true;
    }

    private static long extrapolateNext(long[] seq) {
        List<long[]> sequences = new ArrayList<>();
        sequences.add(seq);

        long[] current = seq;
        while (!allZeros(current)) {
            current = getDifferences(current);
            sequences.add(current);
        }

        // Work backwards from bottom, adding last values
        long extrapolated = 0;
        for (int i = sequences.size() - 2; i >= 0; i--) {
            long[] s = sequences.get(i);
            extrapolated = s[s.length - 1] + extrapolated;
        }

        return extrapolated;
    }

    private static long extrapolatePrev(long[] seq) {
        List<long[]> sequences = new ArrayList<>();
        sequences.add(seq);

        long[] current = seq;
        while (!allZeros(current)) {
            current = getDifferences(current);
            sequences.add(current);
        }

        // Work backwards from bottom, subtracting first values
        long extrapolated = 0;
        for (int i = sequences.size() - 2; i >= 0; i--) {
            long[] s = sequences.get(i);
            extrapolated = s[0] - extrapolated;
        }

        return extrapolated;
    }

    private static long part1(List<long[]> histories) {
        return histories.stream()
                .mapToLong(Solution::extrapolateNext)
                .sum();
    }

    private static long part2(List<long[]> histories) {
        return histories.stream()
                .mapToLong(Solution::extrapolatePrev)
                .sum();
    }
}
